be/src/olap/memtable.cpp

```diff
@@ -115,53 +117,61 @@ int MemTable::RowInBlockComparator::operator()(const RowInBlock* left,
+    {
+        SCOPED_TIMER(_insert_time);
+        if (_is_first_insertion) {
+            _is_first_insertion = false;
+            auto cloneBlock = block->clone_without_columns();
+            _input_mutable_block = vectorized::MutableBlock::build_mutable_block(&cloneBlock);
+            _vec_row_comparator->set_block(&_input_mutable_block);
+            _output_mutable_block = vectorized::MutableBlock::build_mutable_block(&cloneBlock);
+            if (_keys_type != KeysType::DUP_KEYS) {
+                _init_agg_functions(block);
+            }
+        }
+        size_t cursor_in_mutableblock = _input_mutable_block.rows();
+        size_t oldsize = _input_mutable_block.allocated_bytes();
+        _input_mutable_block.add_rows(block, row_pos, num_rows);
+        size_t newsize = _input_mutable_block.allocated_bytes();
+        _mem_usage += newsize - oldsize;
+        _mem_tracker->consume(newsize - oldsize);
+        // when new data inserted, the mem_usage of memtable should be re-shrunk again.
+        _is_shrunk_by_agg = false;
+
+        for (int i = 0; i < num_rows; i++) {
+            _row_in_blocks.emplace_back(new RowInBlock {cursor_in_mutableblock + i});
+            _insert_one_row_from_block(_row_in_blocks.back());
         }
-    }
-    size_t cursor_in_mutableblock = _input_mutable_block.rows();
-    size_t oldsize = _input_mutable_block.allocated_bytes();
-    _input_mutable_block.add_rows(block, row_pos, num_rows);
-    size_t newsize = _input_mutable_block.allocated_bytes();
-    _mem_usage += newsize - oldsize;
-    _mem_tracker->consume(newsize - oldsize);
-
-    for (int i = 0; i < num_rows; i++) {
-        _row_in_blocks.emplace_back(new RowInBlock {cursor_in_mutableblock + i});
-        _insert_one_row_from_block(_row_in_blocks.back());
     }
 }
 
 void MemTable::_insert_one_row_from_block(RowInBlock* row_in_block) {
-    _rows++;
-    bool overwritten = false;
-    if (_keys_type == KeysType::DUP_KEYS) {
-        // TODO: dup keys only need sort opertaion. Rethink skiplist is the beat way to sort columns?
-        _vec_skip_list->Insert(row_in_block, &overwritten);
-        DCHECK(!overwritten) << "Duplicate key model meet overwrite in SkipList";
-        return;
-    }
-
-    bool is_exist = _vec_skip_list->Find(row_in_block, &_vec_hint);
-    if (is_exist) {
-        _aggregate_two_row_in_block(row_in_block, _vec_hint.curr->key);
-    } else {
-        row_in_block->init_agg_places(_agg_functions, _schema->num_key_columns());
-        for (auto cid = _schema->num_key_columns(); cid < _schema->num_columns(); cid++) {
-            auto col_ptr = _input_mutable_block.mutable_columns()[cid].get();
-            auto place = row_in_block->_agg_places[cid];
-            _agg_functions[cid]->add(place,
-                                     const_cast<const doris::vectorized::IColumn**>(&col_ptr),
-                                     row_in_block->_row_pos, nullptr);
+    {
+        SCOPED_TIMER(_sort_agg_time);
```

Comment:
Timers have a certain overhead and I think we should turn them on when we need to debug. And the timing of each line will greatly affect the execution efficiency, we need more efficient but not necessarily accurate timing